BallJoint: apply axis properties to ball joints

srmainwaring opened the pull request:
# 🎉 New feature

Closes #634

## Summary

Allow joint properties to be applied to ball joints in the dart sim physics engine.

As suggested in https://github.com/gazebosim/gz-physics/issues/634, the properties from the first `<axis>` field in the ball joint are applied to all DoF if this element is set. Otherwise the defaults are used, which is current behaviour.

## Test it

- Add an `<axis>` element to a model using a ball joint and set some of the properties (eg. `<limit>`).
- Prior to the change the property would have no effect, after the change the joint range should be restricted.

### Example: `wind.sdf`

- Modify the `wind.sdf` model to include `<axis>` elements for each sphere restricting the movement:

```xml
     <joint name="sphere_1_joint" type="ball">
        <parent>roof</parent>
        <child>sphere_1</child>
        <axis>
          <limit>
            <lower>0</lower>
            <upper>0</upper>
          </limit>
          <dynamics>
            <damping>0.01</damping>
          </dynamics>
        </axis>
      </joint>
 ```

_Figure: Expected `wind.sdf` behaviour with ball joint limit. Wind sock does not move._
![ball_joint_with_limit](https://github.com/user-attachments/assets/5e52458d-51de-4bc2-8d79-c92bdd98eec8)

The motivation for this change is to stabilise a complex closed kinematic chain mechanism employing multiple ball joints (a helicopter rotor head). Without this change the simulation is not stable, and the joints vibrated so badly the simulated model does not maintain integrity (companion change to adjust the dynamic joint erp and cfm properties is also required to get satisfactory behaviour. That is be the subject of the follow up PRs:

- https://github.com/gazebosim/gz-physics/pull/754
- https://github.com/gazebosim/gz-sim/pull/2960

_Figure: Application of ball joint properties to a helicopter rotor head demonstrating swash plate collective and cyclic operation_.

![heli_flapping_blades_1](https://github.com/user-attachments/assets/85622b65-999c-4925-b5d2-9823b8375546)


- [ ] Add test

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers
- [ ] Was GenAI used to generate this PR? If so, make sure to add "Generated-by" to your commits. (See [this policy](https://osralliance.org/wp-content/uploads/2025/05/OSRF-Policy-on-the-Use-of-Generative-Tools-Generative-AI-in-Contributions.pdf) for more info.)

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` and `Generated-by` messages.

ahcorde commented on dartsim/src/SDFFeatures.cc:
```diff
@@ -245,6 +245,36 @@ static dart::dynamics::UniversalJoint *ConstructUniversalJoint(
+  // use default properties if sdfAxis is not set, otherwise apply to all DoF
+  if (sdfAxis)
+  {
+    for (const std::size_t index : {0u, 1u, 2u})
```
include `<cstddef>`

iche033 reviewed (commented):
I added expanded one of the dartsim tests to include loading a ball joint in https://github.com/gazebosim/gz-physics/commit/0cb06bfb55360a54467468a838994cc7fc106cb6. Please take a look and integrate the changes if they look good to you. Feel free to expand the test.

srmainwaring commented:
> I added expanded one of the dartsim tests to include loading a simple ball joint model in [0cb06bf](https://github.com/gazebosim/gz-physics/commit/0cb06bfb55360a54467468a838994cc7fc106cb6). Please take a look and integrate the changes if they look good to you. Feel free to expand the test.

Thanks! LGTM, added in [b6beef1](https://github.com/gazebosim/gz-physics/pull/753/commits/b6beef1cf98a7c5785acc50b321a4c150fbb2326)

srmainwaring commented:
> oh weird. I just pulled in the latest changes and ran the tests locally but everything passed. Need to investigate

I've managed to get the tests running on macOS again (removed a stray conflicting version of `gtest` in `/usr/local`) - and all fine locally - so not sure what's up with CI?

iche033 commented:
As a quick test, I commented out the `<axis>` tag in my branch ([iche033/pr-ball-joint-properties](https://github.com/gazebosim/gz-physics/compare/gz-physics8...iche033/pr-ball-joint-properties)), and [Ubuntu github action](https://github.com/gazebosim/gz-physics/actions/runs/16232856012) passed. For some reason, an extra entity gets added. Still debugging

iche033 commented:
ok finally had more time to look into this. I pushed a change to make CIs green. It was not related to the ball joint axis. I was probably working with outdated branch. The number of entities got bumped to 8 in https://github.com/gazebosim/gz-physics/pull/723 so it makes sense now that we need to bump it again to 9 after adding the model with ball joint.
